Add action handler and action handler capability

liveforeverx opened the pull request:
ActionHandlerCapability is an concept, where describes what action handler
can do. For example: remote execution, execute local command, connect to
ticket system, do something in calendar.

ActionHandler is a remote registration for actual implementation of an
action handler capability. For example: capability is remote execution
and action handler can execute via SSH.

Than, via registration process it will be connected to Configuration node
and implements one or many capabilities (which are connected via implements)

cy303 commented on NTO/Automation/entities/ActionHandlerCapability.ttl:
```diff
@@ -0,0 +1,29 @@
+    a rdfs:Class;
+    rdfs:subClassOf ogit:Entity;
+    rdfs:label "ActionHandlerCapability";
+    dcterms:description """Configuration node to dynamically connect Knowledge Pools or Rankings to a HIRO installation.""";
```
Description should be correct, this is a copy&paste version from configuration node

cy303 commented on NTO/Automation/entities/ActionHandler.ttl:
```diff
@@ -0,0 +1,29 @@
+    a rdfs:Class;
+    rdfs:subClassOf ogit:Entity;
+    rdfs:label "ActionHandlerCapability";
+    dcterms:description """Action handler entity, which should be connected to engine and to multiple .""";
```
Incomplete description?

cy303 commented on NTO/Automation/entities/ActionHandlerCapability.ttl:
```diff
@@ -0,0 +1,29 @@
+    dcterms:creator "dr@arago.de";
+    ogit:scope "NTO";
+    ogit:parent "http://www.purl.org/ogit/Node";
+    ogit:mandatory-attributes (
```
At least fields for the capability name and required/optional parameters should be well defined ogit attributes, we do not want to endorse usage of free attributes all over the place ;)

cy303 commented on NTO/Automation/entities/Configuration.ttl:
```diff
@@ -26,12 +26,13 @@ ogit.Automation:Configuration
 	);
 	ogit:allowed (
+            [ ogit:uses  ogit.Automation:ActionHandler ]
 			[ ogit:uses  ogit.Automation:KIRanking ]
```
shouldn't these be on the same indent level?

liveforeverx replied:
There were spaces instead of tabs, fixed.

ojubran commented on NTO/Automation/attributes/applicability.ttl:
```diff
@@ -0,0 +1,15 @@
+@prefix rdfs:                   <http://www.w3.org/2000/01/rdf-schema#> .
+@prefix dcterms:                <http://purl.org/dc/terms/> .
+
+ogit.Automation:Applicability
```
attributes start with small letters

ojubran commented on NTO/Automation/attributes/applicability.ttl:
```diff
@@ -0,0 +1,15 @@
+ogit.Automation:Applicability
+	a owl:DatatypeProperty;
+	rdfs:subPropertyOf ogit:Attribute;
+	rdfs:label "Applicability";
```
I observed that labels also have small letters

ojubran commented on NTO/Automation/entities/ActionHandlerCapability.ttl:
```diff
@@ -0,0 +1,32 @@
+@prefix rdfs:                   <http://www.w3.org/2000/01/rdf-schema#> .
+@prefix dcterms:                <http://purl.org/dc/terms/> .
+
+ogit.Automation:ActionHandlerCapability
```
I would call it only "Capability"

fotto replied:
or ActionCapability  ?  without "Action" it might to too general

fotto commented on NTO/Automation/attributes/parameters.ttl:
```diff
@@ -0,0 +1,15 @@
+@prefix rdfs:                   <http://www.w3.org/2000/01/rdf-schema#> .
+@prefix dcterms:                <http://purl.org/dc/terms/> .
+
+ogit.Automation:Parameters
```
same here: convert to lowercase

fotto commented on NTO/Automation/entities/ActionHandlerCapability.ttl:
```diff
@@ -0,0 +1,32 @@
+	ogit:scope "NTO";
+	ogit:parent "http://www.purl.org/ogit/Node";
+	ogit:mandatory-attributes (
+		ogit:name
```
add ogit:description?

fotto commented:
general question:  what's the intended usage of ActionHandlerCapability?

a)  shared:  we treat this as a (global) catalogue of capabilities
b) dedicated: a capability is only connected to a single ActionHandler

in case b) we will have lots of definitions of `ExecuteCommand` which could be semantically different
in case a) we would define `ExecuteCommand`  just once. But having the parameter list fixed globally might be too restrictive.

cy303 commented:
a) shared

everything else makes sharing KIs almost impossible if parameters and meanings of different actions are not globally agreed upon. If a specific handler needs custom parameters from the model to perform an action, it might do so using the applicability definition. But different required/optional parameters that need to be set from a KI using the capability should be the same everywhere.

liveforeverx commented:
Fixed. Additionally, changed action handlers to have a map of actionDefinitions, so that action handler represents one instance, but can implement different actions (local, remote, connect to some ticket system and so on).

cy303 commented on NTO/Automation/attributes/actionDefinitions.ttl:
```diff
@@ -0,0 +1,15 @@
+	a owl:DatatypeProperty;
+	rdfs:subPropertyOf ogit:Attribute;
+	rdfs:label "actionDefinitions";
+	dcterms:description "Defines a json map of action handlers with capabilities, applicabilities.";
```
We should try to avoid json encoded blobs in the model if possible, the least we should have here is that the value list has the capability as key and the applicability as value, although I wonder if we could somehow have it "deconstructed" even more... but probably with the applicability value being a map again our possibilities are limited

cy303 commented on NTO/Automation/attributes/parameters.ttl:
```diff
@@ -0,0 +1,15 @@
+	a owl:DatatypeProperty;
+	rdfs:subPropertyOf ogit:Attribute;
+	rdfs:label "parameters";
+	dcterms:description "Contains a parameters in a json map, where key is a parameter name and value is a value of a parameter.";
```
Here I would say we can safely remove the json map reference and just specify key/value semantics in the list values directly

cy303 replied:
one open question here would be how to distinguish between mandatory and optional parameters... could be done either by splitting into "required_parameters" and "optional_parameters" or by defining special values that need to be used for "needs to be specified by user" (simply having empty value could not be distinguished from empty default)

cy303 commented:
has been obsoleted by #447
